Mathlib/Combinatorics/Additive/ErdosGinzburgZiv.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.BigOperators.Ring
+import Mathlib.FieldTheory.ChevalleyWarning
+import Mathlib.RingTheory.UniqueFactorizationDomain
+
+/-!
+# The Erdős–Ginzburg–Ziv theorem
+
+This file proves the Erdős–Ginzburg–Ziv theorem as a corollary of Chevalley-Warning. This theorem
+states that among any (not necessarily distinct) `2 * n - 1` elements of `ZMod n`, we can find `n`
+elements of sum zero.
+
+## Main declarations
+
+* `Int.erdos_ginzburg_ziv`: The Erdős–Ginzburg–Ziv theorem stated using sequences in `ℤ`
+* `ZMod.erdos_ginzburg_ziv`: The Erdős–Ginzburg–Ziv theorem stated using sequences in `ZMod n`
+-/
+
+open Finset MvPolynomial
+open scoped BigOperators
+
+variable {ι : Type*}
+
+section prime
+variable {p : ℕ} [Fact p.Prime] {s : Finset ι}
+
+set_option linter.unusedVariables false in
+/-- The first multivariate polynomial used in the proof of Erdős–Ginzburg–Ziv. -/
+private noncomputable def f₁ (s : Finset ι) (a : ι → ZMod p) : MvPolynomial s (ZMod p) :=
+  ∑ i, X i ^ (p - 1)
+
+/-- The second multivariate polynomial used in the proof of Erdős–Ginzburg–Ziv. -/
+private noncomputable def f₂ (s : Finset ι) (a : ι → ZMod p) : MvPolynomial s (ZMod p) :=
+  ∑ i : s, a i • X i ^ (p - 1)
+
+private lemma totalDegree_f₁_add_totalDegree_f₂ {a : ι → ZMod p} :
+    (f₁ s a).totalDegree + (f₂ s a).totalDegree < 2 * p - 1 := by
+  refine (add_le_add (totalDegree_finset_sum _ _) $ (totalDegree_finset_sum _ _).trans $
+    Finset.sup_mono_fun fun a _ ↦ totalDegree_smul_le _ _).trans_lt ?_
+  simp only [totalDegree_X_pow, ← two_mul]
+  refine (mul_le_mul_left' Finset.sup_const_le _).trans_lt ?_
+  rw [mul_tsub, mul_one]
+  exact tsub_lt_tsub_left_of_le ((Fact.out : p.Prime).two_le.trans $
```

Comment:
Isn't there an instance from Fact Prime to AtLeastTwo which can help you here? If not, how about using `Prime.one_lt'` instead?

Reply:
I don't think I should be relying on obscure machinery for such simple statements. I also don't see how `(Nat.Prime;one_lt' p).out` is better than `(Fact.out : p.Prime).two_le`